### PR TITLE
[fix] [Perf] PerformanceProducer do not produce expected number of messages.

### LINKED ISSUE #19774: [Bug] [Perf] PerformanceProducer do not produce expected number of messages.
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

2.9.

### Minimal reproduce step

Use PerformanceProducer with `--num-messages` to determine the number of messages that i want to produce, and with `--num-test-threads` to start multiple threads to working.
```
bin/pulsar-perf produce --num-messages 10000000 --num-test-threads 4 ...
```

### What did you expect to see?

I want to produce exactly 100000 messages to be produced.

### What did you see instead?

Less than 10000000 messages are produced.

### Anything else?

The root reason is as follows:
When we use configuration `--num-messages 100000 --num-test-threads 4` to start perf process, process will divide the total messages with threads count, that is `100000/4=25000`, the result of which will be assign as the total messages to be produced by each threads. 
So, each thread need to 25000 messages.
But, when one of threads produce 25000 messages, it will call `PerfClientUtils.exit(0);` to terminates the currently running Java Virtual Machine, which will result into that other threads is terminated but do not produce enough messages yet.
image

The test code bellow can simulate this process:
```
    public static void main(String[] args) throws InterruptedException {
        ExecutorService executor = Executors
                .newCachedThreadPool(new DefaultThreadFactory("pulsar-perf-producer-exec"));
        CountDownLatch doneLatch = new CountDownLatch(2);
        executor.submit(()->{
            try {
                Thread.sleep(10000);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            System.out.println("thread1: need to be printed.");
            doneLatch.countDown();
        });
        executor.submit(()->{
            System.out.println("thread2: call system.exit(0).");
            doneLatch.countDown();
            System.exit(0);
        });
        while (true) {
            if (doneLatch.getCount() <= 0) {
                break;
            }
            Thread.sleep(10000);
        }
    }
```
the output as follows:
image

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by thetumbled**
Fixes #19774 

### Motivation
when one of threads produce enough messages, it will shutdown all threads, which may do not produce enough messages yet.

### Modifications
remove `system.exit()`.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/thetumbled/pulsar/pull/19

### CODE REVIEW

File: pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java [hunk 1]
```diff
@@ -604,14 +606,18 @@ private static void runProducer(int producerId,
             AtomicLong numMessageSend = new AtomicLong(0);
             Semaphore numMsgPerTxnLimit = new Semaphore(arguments.numMessagesPerTransaction);
             while (true) {
+                if (produceEnough) {
```
Comment (AnonHxy) [inline pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java:609]: Can `produceEnough` be true here?

File: pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java [hunk 2]
```diff
@@ -753,11 +760,13 @@ private static void runProducer(int producerId,
             if (null != client) {
                 try {
                     client.close();
-                    PerfClientUtils.exit(1);
                 } catch (PulsarClientException e) {
                     log.error("Failed to close test client", e);
                 }
             }
+            if (!produceEnough) {
```
Comment (AnonHxy) [inline pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java:767]: How about move this block to line before `if (null != client)`?  Or catch `Throwable` when client close.

File: pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java (hunk 1 above)
Comment (thetumbled) [reply pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java:609]: If we produce enough messages or produce for enough time, `produceEnough` will be set to `true`.

File: pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java (hunk 2 above)
Comment (thetumbled) [reply pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java:767]: why do we move this block to line before if (null != client)?

File: pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java (hunk 1 above)
Comment (AnonHxy) [reply pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java:609]: OK, I missed the inner-for loop

File: pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java (hunk 2 above)
Comment (AnonHxy) [reply pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java:767]: Could `client.close` throw other exceptions , which will cause this block not be executed

Comment (thetumbled) [reply pulsar-testclient/src/main/java/org/apache/pulsar/testclient/PerformanceProducer.java]: you are right, i have fixed it.

**Review by AnonHxy — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

**Review by tisonkun — APPROVED** [review]
Thank you! Merging...

**Comment by michaeljmarshall** [comment]
As discussed on the mailing list https://lists.apache.org/thread/w4jzk27qhtosgsz7l9bmhf1t7o9mxjhp, there is no plan to release 2.9.6, so I am going to remove the release/2.9.6 label

**Pull request merged**